dozer-core/src/checkpoint/mod.rs

```diff
@@ -81,19 +151,64 @@ impl CheckpointFactory {
         &self.record_store
     }
 
-    fn write_record_store_slice(&self, key: String) -> Result<(), ExecutionError> {
+    fn write_record_store_slice(
+        &self,
+        key: String,
+        source_states: &SourceStates,
+    ) -> Result<(), ExecutionError> {
         let mut state = self.state.lock();
         let (data, num_records_serialized) =
             self.record_store.serialize_slice(state.next_record_index)?;
         state.next_record_index += num_records_serialized;
         drop(state);
 
-        self.queue
-            .upload_object(key, data)
-            .map_err(|_| ExecutionError::CheckpointWriterThreadPanicked)?;
+        self.write_record_store_slice_data(key, source_states, data)
+            .map_err(|_| ExecutionError::CheckpointWriterThreadPanicked)
+    }
+
+    fn write_record_store_slice_data(
```

Comment:
I didn't want to clone `data` during deserialization, which can be quite big. Not an expert with `serde` so don't know how to do that. Any quick pointers?

Reply:
In this case there would, I think, be no way to avoid a clone of the data, as bincode at least has to sequence all of it together into a single `Vec<u8>`.